File: quant1x/net/connection_pool.py
"""Python implementation of TcpConnectionPool mirroring the C++ design.

This module implements:
- Connection: RAII-like wrapper around a socket (life-cycle owned by the pool)
- TcpConnectionPool: thread-safe pool matching the C++ API and semantics

Notes:
- Endpoints are (host, port) tuples and managed by EndpointManager from
  quant1x.net.endpoint.
- Network handler is a user-provided object exposing the methods:
  - timeout() -> float (seconds)
  - handshake(socket.socket) -> bool
  - keepalive(socket.socket) -> bool
  - check_interval() -> float (seconds)

The pool returns a ConnectionHandle which is a context-manager and will
automatically return the connection to the pool when exiting the context or
when the handle is garbage collected.
"""
from __future__ import annotations

import socket
import threading
import time
import logging
from collections import deque
from typing import Optional, Callable, Deque, Tuple

from .endpoint import EndpointManager, Endpoint

log = logging.getLogger(__name__)


class Connection:
    """Wrapper for a connected socket and its endpoint.

    Lifecycle is managed by TcpConnectionPool; do not close externally unless
    explicitly desired.
    """

    def __init__(self, sock: socket.socket, endpoint: Endpoint):
        if sock is None or sock.fileno() == -1:
            raise ValueError("Socket must be connected")
        self._sock = sock
        self._endpoint = endpoint

    @property
    def socket(self) -> socket.socket:
        return self._sock

    @property
    def endpoint(self) -> Endpoint:
        return self._endpoint

    def close(self) -> None:
        try:
            try:
                self._sock.shutdown(socket.SHUT_RDWR)
            except Exception:
                pass
            self._sock.close()
        finally:
            return

    def is_open(self) -> bool:
        try:
            return self._sock.fileno() != -1
        except Exception:
            return False


class ConnectionHandle:
    """A RAII-style handle returned by TcpConnectionPool.acquire().

    The handle can be used as a context manager. When closed/released it will
    return the Connection object to the originating pool.
    """

    def __init__(self, conn: Connection, releaser: Callable[[Connection], None]):
        self._conn = conn
        self._releaser = releaser
        self._released = False

    def __enter__(self) -> Connection:
        return self._conn

    def __exit__(self, exc_type, exc, tb):
        self.release()

    def release(self) -> None:
        if not self._released:
            try:
                self._releaser(self._conn)
            finally:
                self._released = True

    def __del__(self):
        # best-effort auto-release to mimic C++ unique_ptr deleter
        try:
            self.release()
        except Exception:
            log.exception("Error during auto-release of connection")


class TcpConnectionPool:
    def __init__(self, min_connections: int, max_connections: int, network_handler):
        if min_connections > max_connections:
            raise ValueError("min_connections cannot be greater than max_connections")
        if max_connections == 0:
            raise ValueError("max_connections cannot be zero")
        if network_handler is None:
            raise ValueError("network_handler cannot be None")

        self.min_connections = int(min_connections)
        self.max_connections = int(max_connections)
        self.endpoint_weight = 1
        self.network_handler = network_handler
        self.endpoint_manager = EndpointManager()

        self._connections_mutex = threading.Lock()
        self._idle_connections: Deque[Connection] = deque()
        self.idle_connection_count = 0
        self.active_connection_count = 0

        self.running = False
        self._heartbeat_timer: Optional[threading.Timer] = None

        # Start background thread executing heartbeat and creation attempts
        self._worker_threads = []
        # create 2 worker threads to mimic C++ io threads (they only run timers here)
        for i in range(2):
            t = threading.Thread(target=self._worker_loop, name=f"connpool-io-{i}", daemon=True)
            self._worker_threads.append(t)
            t.start()

        # start pool
        self.start()

    # ---------- endpoint helpers (C++-compatible names) ----------
    def add_endpoint(self, ip: str, port: int, weight: int = 0) -> bool:
        return self.endpoint_manager.add_endpoint(ip, port, weight or self.endpoint_weight)

    def addEndpoint(self, ip: str, port: int, weight: int = 0) -> bool:
        return self.add_endpoint(ip, port, weight)

    def add_endpoint_obj(self, endpoint: Endpoint, weight: int = 0) -> bool:
        return self.endpoint_manager.add_endpoint_obj(endpoint, weight or self.endpoint_weight)

    # ---------- core pool API ----------
    def acquire(self) -> ConnectionHandle:
        # 1. try reuse
        raw_conn: Optional[Connection] = None
        with self._connections_mutex:
            if self._idle_connections:
                raw_conn = self._idle_connections.popleft()
                self.idle_connection_count -= 1
                log.debug("Reused connection from pool: %s", raw_conn)

        # 2. if none, create new
        endpoint = None
        if raw_conn is None:
            log.debug("Creating new connection...")
            ep = self.endpoint_manager.acquire_endpoint()
            if ep is None:
                log.error("No available endpoints")
                raise RuntimeError("No available endpoints")
            endpoint = ep
            sock = None
            try:
                # create and connect with timeout from handler
                timeout = float(self.network_handler.timeout())
                sock = socket.create_connection(endpoint, timeout=timeout)
                # also set socket timeout for subsequent recv/send operations so
                # blocking reads used by protocol helpers (process_request_std)
                # will raise socket.timeout instead of blocking forever.
                try:
                    sock.settimeout(timeout)
                except Exception:
                    pass
                # set TCP_NODELAY
                try:
                    sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
                except Exception:
                    pass

                # handshake
                ok = self.network_handler.handshake(sock)
                if not ok:
                    sock.close()
                    self.endpoint_manager.release_endpoint(endpoint)
                    log.error("Handshake failed with %s:%s", endpoint[0], endpoint[1])
                    raise RuntimeError("Handshake failed")

                raw_conn = Connection(sock, endpoint)
                log.debug("Created new connection %s", raw_conn)
            except Exception:
                if sock is not None:
                    try:
                        sock.close()
                    except Exception:
                        pass
                # release endpoint reservation
                if endpoint is not None:
                    self.endpoint_manager.release_endpoint(endpoint)
                raise

        # create deleter/releaser
        def deleter(conn: Connection) -> None:
            try:
                self.release(conn)
            except Exception:
                log.exception("Error during auto-release")

        self.active_connection_count += 1
        return ConnectionHandle(raw_conn, deleter)

    def release(self, conn: Connection) -> None:
        if conn is None:
            return
        conn_id = id(conn)
        log.debug("Returning connection %s", conn_id)

        # Do NOT release endpoint here (same semantics as C++)
        with self._connections_mutex:
            self._idle_connections.append(conn)
            self.idle_connection_count += 1
            self.active_connection_count -= 1
            log.debug("Connection %s returned to pool", conn_id)

    def close_connection(self, conn: Optional[Connection]) -> None:
        if conn is None:
            return
        conn_id = id(conn)
        log.debug("Closing connection %s", conn_id)
        # 1. release endpoint
        try:
            self.endpoint_manager.release_endpoint(conn.endpoint)
        except Exception:
            log.exception("Error releasing endpoint")
        # 2. close socket
        try:
            conn.close()
        except Exception:
            log.exception("Error closing connection socket")
        # 3. update counters
        with self._connections_mutex:
            self.idle_connection_count = max(0, self.idle_connection_count - 1)

    def start(self) -> None:
        if self.running:
            return
        self.running = True
        self._start_heartbeat_timer()

    def stop(self) -> None:
        self.running = False
        if self._heartbeat_timer:
            self._heartbeat_timer.cancel()
        self._close_all_connections()

    def get_endpoint_stats(self, host: str, port: int):
        ep = (host, int(port))
        try:
            return self.endpoint_manager.get_endpoint_stats(ep)
        except KeyError as e:
            raise RuntimeError("Invalid endpoint: %s" % e)

    # ---------- internal helpers ----------
    def _start_heartbeat_timer(self) -> None:
        interval = float(self.network_handler.check_interval())

        def tick():
            if not self.running:
                return
            try:
                self._check_connections()
                self._try_create_connections()
            except Exception:
                log.exception("Exception in heartbeat tick")
            # reschedule
            if self.running:
                self._heartbeat_timer = threading.Timer(interval, tick)
                self._heartbeat_timer.daemon = True
                self._heartbeat_timer.start()

        self._heartbeat_timer = threading.Timer(interval, tick)
        self._heartbeat_timer.daemon = True
        self._heartbeat_timer.start()

    def _check_connections(self) -> None:
        with self._connections_mutex:
            new_idle = deque()
            while self._idle_connections:
                conn = self._idle_connections.popleft()
                if conn is None:
                    continue
                try:
                    alive = self.network_handler.keepalive(conn.socket)
                except Exception:
                    # treat exception as dead
                    self.close_connection(conn)
                    continue
                if not alive:
                    self.close_connection(conn)
                else:
                    new_idle.append(conn)
            self._idle_connections = new_idle

    def _try_create_connections(self) -> None:
        max_retries = 10
        retry_count = 0
        while self.active_connection_count + self.idle_connection_count < self.min_connections and retry_count < max_retries:
            available = self.endpoint_manager.get_available_resources()
            if available == 0:
                log.warning("endpoint resources exhausted, retry %s/%s", retry_count, max_retries)
                time.sleep(0.1)
                retry_count += 1
                continue
            else:
                try:
                    handle = self.acquire()
                    # release immediately to put into idle pool
                    handle.release()
                    retry_count += 1
                    log.debug("Supplemented 1 connection")
                except Exception:
                    log.exception("Error acquiring new connection")
                    break

    def _close_all_connections(self) -> None:
        with self._connections_mutex:
            for conn in list(self._idle_connections):
                try:
                    if conn and conn.is_open():
                        conn.close()
                except Exception:
                    log.exception("Error closing idle connection")
            self._idle_connections.clear()
            self.idle_connection_count = 0

    def _worker_loop(self) -> None:
        # simple loop to keep threads alive; the real work is done by timers
        while self.running:
            time.sleep(0.5)

File: quant1x/net/connection_pool.go
package net

import (
	"context"
	"errors"
	"fmt"
	stdnet "net"
	"sync"
	"time"

	"gitee.com/quant1x/quant1x/quant1x/log"
)

// Connection is an RAII-style wrapper around an established TCP connection.
// Its lifecycle is managed by TcpConnectionPool; callers should not copy it.
type Connection struct {
	conn     *stdnet.TCPConn
	endpoint *stdnet.TCPAddr
}

func NewConnection(c *stdnet.TCPConn, ep *stdnet.TCPAddr) *Connection {
	return &Connection{conn: c, endpoint: ep}
}

func (c *Connection) Conn() *stdnet.TCPConn     { return c.conn }
func (c *Connection) Endpoint() *stdnet.TCPAddr { return c.endpoint }

func (c *Connection) Close() {
	if c == nil || c.conn == nil {
		return
	}
	_ = c.conn.Close()
}

func (c *Connection) IsOpen() bool {
	return c != nil && c.conn != nil
}

// TcpConnectionPool manages a pool of connections across multiple endpoints.
type TcpConnectionPool struct {
	// configuration
	minConnections int
	maxConnections int
	endpointWeight int

	// dependencies
	networkHandler  NetworkOperationHandler
	endpointManager *EndpointManager

	// lifecycle
	ctx     context.Context
	cancel  context.CancelFunc
	running bool

	// connection storage
	connectionsMutex sync.Mutex
	idleConnections  []*Connection
	idleCount        int
	activeCount      int

	// heartbeat
	heartbeatTicker *time.Ticker
}

// NewTcpConnectionPool creates a new pool. networkHandler must be non-nil.
func NewTcpConnectionPool(minConn, maxConn int, handler NetworkOperationHandler) (*TcpConnectionPool, error) {
	if minConn > maxConn {
		return nil, fmt.Errorf("min_connections cannot be greater than max_connections")
	}
	if maxConn == 0 {
		return nil, fmt.Errorf("max_connections cannot be zero")
	}
	if handler == nil {
		return nil, fmt.Errorf("network_handler cannot be nil")
	}

	ctx, cancel := context.WithCancel(context.Background())
	p := &TcpConnectionPool{
		minConnections:  minConn,
		maxConnections:  maxConn,
		endpointWeight:  1,
		networkHandler:  handler,
		endpointManager: NewEndpointManager(),
		ctx:             ctx,
		cancel:          cancel,
	}

	log.Debugf("[connection pool] min_connections=%d, max_connections=%d, endpoint_weight=%d", minConn, maxConn, p.endpointWeight)

	// start background heartbeat
	p.start()
	return p, nil
}

// AddEndpoint adds an endpoint by host/port
func (p *TcpConnectionPool) AddEndpoint(host string, port int, weight int) bool {
	if weight == 0 {
		weight = p.endpointWeight
	}
	return p.endpointManager.AddEndpoint(host, uint16(port), weight)
}

// AddEndpointAddr adds an endpoint by *net.TCPAddr (address object)
func (p *TcpConnectionPool) AddEndpointAddr(addr *stdnet.TCPAddr, weight int) bool {
	if weight == 0 {
		weight = p.endpointWeight
	}
	return p.endpointManager.AddEndpointAddr(addr, weight)
}

// Acquire attempts to reuse an idle connection or create a new one. It returns the
// connection and a release function which must be called when done with the connection.
func (p *TcpConnectionPool) Acquire() (*Connection, func(), error) {
	// 1. try reuse
	var c *Connection
	p.connectionsMutex.Lock()
	if len(p.idleConnections) > 0 {
		c = p.idleConnections[len(p.idleConnections)-1]
		p.idleConnections = p.idleConnections[:len(p.idleConnections)-1]
		p.idleCount--
		log.Debugf("Reused connection from pool (ptr: %p)", c)
	}
	p.connectionsMutex.Unlock()

	// 2. create new if needed
	if c == nil {
		log.Debugf("Creating new connection...")
		ep, ok := p.endpointManager.AcquireEndpoint()
		if !ok || ep == nil {
			log.Errorf("No available endpoints")
			return nil, nil, errors.New("no available endpoints")
		}

		// Dial with timeout
		dialer := &stdnet.Dialer{Timeout: p.networkHandler.Timeout()}
		addr := ep.String()
		rawConn, err := dialer.DialContext(p.ctx, "tcp", addr)
		if err != nil {
			p.endpointManager.ReleaseEndpoint(ep)
			log.Errorf("Error dialing %s: %v", addr, err)
			return nil, nil, err
		}
		tcpConn, ok2 := rawConn.(*stdnet.TCPConn)
		if !ok2 {
			rawConn.Close()
			p.endpointManager.ReleaseEndpoint(ep)
			return nil, nil, fmt.Errorf("unexpected connection type")
		}

		// handshake
		okHandshake, err := p.networkHandler.Handshake(tcpConn)
		if err != nil || !okHandshake {
			tcpConn.Close()
			p.endpointManager.ReleaseEndpoint(ep)
			log.Errorf("Handshake failed with %s: %v", addr, err)
			if err == nil {
				err = errors.New("handshake failed")
			}
			return nil, nil, err
		}

		c = NewConnection(tcpConn, ep)
		log.Debugf("Created new connection (ptr: %p)", c)
	}

	p.connectionsMutex.Lock()
	p.activeCount++
	p.connectionsMutex.Unlock()

	// release function
	released := false
	releaseFunc := func() {
		if released {
			return
		}
		released = true
		p.Release(c)
	}
	return c, releaseFunc, nil
}

// Release returns a connection to the idle pool
func (p *TcpConnectionPool) Release(c *Connection) {
	if c == nil {
		return
	}
	connPtr := fmt.Sprintf("%p", c)
	log.Debugf("Returning connection %s to pool", connPtr)

	p.connectionsMutex.Lock()
	defer p.connectionsMutex.Unlock()
	p.idleConnections = append(p.idleConnections, c)
	p.idleCount++
	if p.activeCount > 0 {
		p.activeCount--
	}
}

// closeConnection closes a single idle connection and releases its endpoint
func (p *TcpConnectionPool) closeConnection(c *Connection) {
	if c == nil {
		return
	}
	log.Debugf("Closing connection (ptr: %p)", c)
	// Release the endpoint back to manager, then close the socket.
	// This mirrors the C++ closeConnection behavior which returns the
	// endpoint allocation when a connection is being permanently closed.
	p.endpointManager.ReleaseEndpoint(c.Endpoint())
	c.Close()

	p.connectionsMutex.Lock()
	if p.idleCount > 0 {
		p.idleCount--
	}
	p.connectionsMutex.Unlock()
}

// start launches heartbeat timer and background routines
func (p *TcpConnectionPool) start() {
	if p.running {
		return
	}
	p.running = true
	interval := p.networkHandler.CheckInterval()
	if interval <= 0 {
		interval = 5 * time.Second
	}
	p.heartbeatTicker = time.NewTicker(interval)
	go func() {
		for {
			select {
			case <-p.ctx.Done():
				log.Infof("heartbeat exiting")
				return
			case <-p.heartbeatTicker.C:
				if !p.running {
					log.Infof("heartbeat stopping")
					return
				}
				p.checkConnections()
				p.tryCreateConnections()
			}
		}
	}()
}

// Stop stops the pool and closes connections
func (p *TcpConnectionPool) Stop() {
	if !p.running {
		return
	}
	p.running = false
	if p.heartbeatTicker != nil {
		p.heartbeatTicker.Stop()
	}
	p.cancel()
	p.closeAllConnections()
}

// checkConnections validates idle connections using handler.Keepalive
func (p *TcpConnectionPool) checkConnections() {
	p.connectionsMutex.Lock()
	defer p.connectionsMutex.Unlock()
	// Follow C++ semantics closely:
	// - If keepalive() returns true, keep the connection in idle pool.
	// - If keepalive() returns false, the connection is removed and the
	//   socket is closed (the endpoint allocation is NOT released here),
	//   mirroring the C++ remove_if + unique_ptr destruction behavior.
	// - If keepalive() throws/returns an error, we call closeConnection
	//   which closes the socket AND releases the endpoint allocation.

	var survivors []*Connection
	for _, conn := range p.idleConnections {
		if conn == nil {
			continue
		}
		ok, err := p.networkHandler.Keepalive(conn.Conn())
		if err != nil {
			// Exception path: permanently close and release endpoint
			log.Errorf("keepalive error: %v", err)
			p.connectionsMutex.Unlock() // unlock while performing closeConnection which will lock
			p.closeConnection(conn)
			p.connectionsMutex.Lock()
			continue
		}
		if ok {
			survivors = append(survivors, conn)
		} else {
			// keepalive=false: close socket but DO NOT release endpoint here
			log.Debugf("connection not healthy, closing socket: %p", conn)
			conn.Close()
			// do NOT call ReleaseEndpoint here (mirror C++ behavior)
		}
	}
	p.idleConnections = survivors
	p.idleCount = len(survivors)
}

// tryCreateConnections attempts to ensure minConnections are available
func (p *TcpConnectionPool) tryCreateConnections() {
	retries := 0
	for p.activeCount+p.idleCount < p.minConnections && retries < 10 {
		available := p.endpointManager.GetAvailableResources()
		if available == 0 {
			log.Infof("endpoint resources insufficient, retry %d/10", retries)
			time.Sleep(100 * time.Millisecond)
			retries++
			continue
		}
		conn, release, err := p.Acquire()
		if err != nil {
			log.Errorf("Error acquiring new connection: %v", err)
			break
		}
		// Immediately release the connection to mirror C++ unique_ptr
		// lifetime behavior in try_create_connections (destructor returns
		// connection to pool).
		if release != nil {
			release()
		}
		// increment retry counter on attempt
		retries++
		log.Debugf("supplemented 1 connection, endpoint=%v", conn.Endpoint())
	}
}

// closeAllConnections closes all idle connections
func (p *TcpConnectionPool) closeAllConnections() {
	p.connectionsMutex.Lock()
	defer p.connectionsMutex.Unlock()
	for _, c := range p.idleConnections {
		if c != nil && c.IsOpen() {
			c.Close()
		}
	}
	p.idleConnections = nil
	p.idleCount = 0
}

// GetEndpointStats returns stats for a given host/port
func (p *TcpConnectionPool) GetEndpointStats(host string, port int) (int, int, error) {
	ep, err := stdnet.ResolveIPAddr("ip", host)
	if err != nil {
		return 0, 0, err
	}
	tcpAddr := &stdnet.TCPAddr{IP: ep.IP, Port: port}
	max, active, err := p.endpointManager.GetEndpointStats(tcpAddr)
	if err != nil {
		return 0, 0, err
	}
	return int(max), int(active), nil
}
